src/Microsoft.Azure.WebJobs.Host/Executors/FunctionExecutor.cs

```diff
@@ -175,6 +175,12 @@ private async Task<IDelayedException> TryExecuteAsyncCore(IFunctionInstanceEx fu
                 await HandleExceptionAsync(functionInstance.FunctionDescriptor.TimeoutAttribute, exceptionInfo, _exceptionHandler);
             }
 
+            if (parameterHelper.ReturnValue is RetryResult retry)
+            {
+                // Set RetryException here so the execution is not logged as failed
+                exceptionInfo = ExceptionDispatchInfo.Capture(new RetryException(retry));
```

Comment:
I don't understand why we need the RetryException/RetryResult types at all. It seems RetryManager can just handle retries internally, handling all exceptions and state internally. If the final retry fails, it just lets the error bubble out naturally